File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/JsonBase64DataSerializer.java
package com.github.dedis.popstellar.model.network.serializer;

import com.github.dedis.popstellar.model.objects.security.Base64URLData;
import com.google.gson.*;

import java.lang.reflect.Type;
import java.util.function.Function;

public class JsonBase64DataSerializer<T extends Base64URLData>
    implements JsonSerializer<T>, JsonDeserializer<T> {

  private final Function<String, T> constructor;

  public JsonBase64DataSerializer(Function<String, T> constructor) {
    this.constructor = constructor;
  }

  @Override
  public T deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
      throws JsonParseException {
    try {
      return constructor.apply(json.getAsString());
    } catch (Exception e) {
      throw new JsonParseException(e);
    }
  }

  @Override
  public JsonElement serialize(T src, Type typeOfSrc, JsonSerializationContext context) {
    return new JsonPrimitive(src.getEncoded());
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/serializer/base64/JsonBase64DataSerializer.kt
package com.github.dedis.popstellar.model.network.serializer.base64

import com.github.dedis.popstellar.model.objects.security.Base64URLData
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import java.lang.reflect.Type
import java.util.function.Function

class JsonBase64DataSerializer<T : Base64URLData?>(private val constructor: Function<String, T>) : JsonSerializer<T>, JsonDeserializer<T> {
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): T {
        return try {
            constructor.apply(json.asString)
        } catch (e: Exception) {
            throw JsonParseException(e)
        }
    }

    override fun serialize(src: T, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        return JsonPrimitive(src!!.encoded)
    }
}